src/nvme/core.c

```diff
@@ -49,6 +50,66 @@
 #define sqtdbl(doorbells, qid, dstrd) \
 	(doorbells + (2 * qid) * (4 << dstrd))
 
+struct nvme_ctrl_handle {
+	struct nvme_ctrl *ctrl;
+	struct list_node list;
+};
+
+static LIST_HEAD(nvme_ctrl_handles);
+
+static struct nvme_ctrl_handle *nvme_get_ctrl_handle(const char *bdf)
+{
+	struct nvme_ctrl_handle *handle;
+
+	list_for_each(&nvme_ctrl_handles, handle, list) {
+		if (streq(handle->ctrl->pci.bdf, bdf))
+			return handle;
+	}
+
+	return NULL;
+}
+
+
+struct nvme_ctrl *nvme_get_ctrl(const char *bdf)
+{
+	struct nvme_ctrl_handle *handle = nvme_get_ctrl_handle(bdf);
+
+	if (!handle)
+		return NULL;
+
+	return handle->ctrl;
+}
+
+int nvme_add_ctrl(struct nvme_ctrl *ctrl)
+{
+	struct nvme_ctrl_handle *handle;
+
+	if (nvme_get_ctrl(ctrl->pci.bdf)) {
+		errno = EALREADY;
```

Comment:
`EALREADY` is usually used for network, no? Not that `EEXIST` is that much better (referring to files), but we already use that in other places.